targets/TARGET_Analog_Devices/TARGET_ADUCM302X/TARGET_ADUCM3029/api/trng_api.c

```diff
@@ -0,0 +1,124 @@
+ *   - Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *   - Modified versions of the software must be conspicuously marked as such.
+ *   - This software is licensed solely and exclusively for use with processors
+ *     manufactured by or for Analog Devices, Inc.
+ *   - This software may not be combined or merged with other code in any manner
+ *     that would cause the software to become subject to terms and conditions
+ *     which differ from those listed here.
+ *   - Neither the name of Analog Devices, Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *   - The use of this software may or may not infringe the patent rights of one
+ *     or more patent holders.  This license does not release you from the
+ *     requirement that you obtain separate licenses from these patent holders
+ *     to use this software.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES, INC. AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-
+ * INFRINGEMENT, TITLE, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ANALOG DEVICES, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, DAMAGES ARISING OUT OF
+ * CLAIMS OF INTELLECTUAL PROPERTY RIGHTS INFRINGEMENT; PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#if defined(DEVICE_TRNG)
+
+#include <stdlib.h>
+#include <drivers/rng/adi_rng.h>
+#include <drivers/pwr/adi_pwr.h>
+#include "cmsis.h"
+#include "trng_api.h"
+#include "objects.h"
+
+// Sampling counter values
+// Prescaler: 0 - 10
+// LenReload: 0 - 4095
+#define TRNG_CNT_VAL    4095
+#define TRNG_PRESCALER  2
+
+/* RNG Device memory */
+static uint8_t RngDevMem[ADI_RNG_MEMORY_SIZE];
+
+void trng_init(trng_t *obj)
+{
+    ADI_RNG_HANDLE RNGhDevice;
+
+    // Open the device
+    adi_rng_Open(0,RngDevMem,sizeof(RngDevMem),&RNGhDevice);
+
+    // Set sample length for the H/W RN accumulator
+    adi_rng_SetSampleLen(RNGhDevice, TRNG_PRESCALER, TRNG_CNT_VAL);
+
+    // Disable buffering - single byte generation only
+    adi_rng_EnableBuffering(RNGhDevice, false);
+
+    // Enable the TRNG
+    adi_rng_Enable(RNGhDevice, true);
```

Comment:
```adi_rng_EnableBuffering``` and ```adi_rng_Enable``` can fail based on definition in ```adi_rng.h```. Wouldn't it be good idea to check status here. Also, initialise ```memset( (void *)obj, 0, sizeof( trng_t ) );``` so that uninitialised object can be checked in other functions using the ```obj```.

Reply:
@mazimkhan That's a general problem with the interfaces I'd say, since `trng_init` is not allowed to fail (and there's no `xxx_setup` function as in other modules). An ugly way to get around this would be to extend
~~~
struct trng_s {
    ADI_RNG_HANDLE  RNGhDevice;
};
~~~
by some flag indicating that something went wrong during initialization, and checking for that in `trng_get_bytes`.

Reply:
@hanno-arm But if the adi_xxx functions fail, we should assertr in this case, or something of the sort. I don't think we should continue operation

Reply:
@RonEld I agree, but the interface doesn't allow to signal the error. I don't know if an `assert` would be acceptable here and how it would affect overall system operation @0xc0170.

Reply:
@dave-wu do you agree to `error()` if init fails ? If yes, please add it as requested above